### ISSUE TITLE
[feature] system_requirements generator

### ISSUE THREAD

**Issue by KerstinKeller** [conan-io/conan#8451]
After a discussion on Slack with @solvingj , I am raising an issue here on Github.

It would be great to have some kind of `system_requirements` generator, that, in the easiest case, just prints a text file which contains all system requirements needed to build / run a specific Conan package.

We've setup conan such that system packages won't be installed silently, but it would be awesome to check if everything is in place separately, and most importantly deploy a list of the system requirements together with the rest. 

It seems that there is no way to achieve this as of now. So far I don't see any output of system requirements in any of the files produced by a `conan install`.

From the discussion I've taken that I could manually add the information to `self.user_info`? I can try and implement a generator by using the workaround, but I guess that that information will be important for other users, as well.

- [x ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by solvingj** [comment]
This is a great suggestion. 

A few thoughts for future reference about small-scope incremental improvements for the general case of `system_requirements`: 

-  The simplest enhancement which could be made around `system_requirements` today would be to add additional modes for the configuration item of `CONAN_SYSREQUIRES_MODE`.  One mode that comes to mind after hearing your case would be something like `warn` / `print` / `show` / `suggest` / `log`.  Each of these words would imply a slightly different behavior, and it's not clear which would be best, but the key characteristics for a new mode would be: "non-intrusive", "non-fatal", and to produce "machine-readable results".  It's possible we may want to promote this as an independent feature request. 

- Independently of new modes, the new `[conf]` mechanism seems like a good fit to enhance the `system_requirements()` in other ways. For example, we already know we want to eventually deprecate many environment variables like `CONAN_SYSREQUIRES_MODE`.  `[conf]` provides an alternative which offers many advantages out of the box.  
 
First, users can specify behavior per-package, and eventually from the CLI, with familiar syntax:
```
    --conf some_package:sysrequires_mode=log   --conf *:sysrequires_mode=disabled
```
Second, it would enable additional parameters. Previously, we had one related parameter of `CONAN_SYSREQUIRES_SUDO`. But this strategy doesn't scale for more parameters.  With `[conf]`, we can support more interesting parameters in a more manageable way like this:
```
    --conf package_name:sysrequires_outfile=/path/to/my_pkg_sysrequires.log
```

Now, here are some thoughts about a `system_requirements` generator. 

- You mentioned a simple text list.  Yes, this is probably the best foundation.  But, similar to the `virtualenv` generators, it seems likely this would be much cooler to generate accompanying scripts which read the text files and actually run the installs accordingly.  

- What fundamental additions or changes would need to happen for this?  

- First, the packages would need to be able to advertise the list via `package_info`.  There's no data structure available for this info, so it would probably require adding `sysrequires_info` as neighbors to `cpp_info`, `user_info`, and `env_info`.  That seems... easily to do, if it's found to be technically agreeable. 

- After that, we need just need to assign the list (or maybe a dictionary) of package names as the value to that new variable, and then a generator could process the data and output it however is needed.  The problem is that currently, existing recipes pass each item which they want to install as parameters in a bunch of separate function calls: something like  `installer.install(package_name)`.  So, existing recipes would need to be refactored, to make that data separate from the behavior so it could be gathered in `system_requirements()` method and `package_info()` method. That's a bummer. 

- There's an interesting "hack" which might make such a thing possible without refactoring recipes. I don't advise it, but it's worth pitching.  Here's code you might find in a recipe today:
```
        installer = SystemPackageTool(tool=ChocolateyTool()) # Invoke choco package manager to install the package
        installer.install(pack_name)
```
`SystemPackageTool` is a class we control.  We COULD (probably) implicitly call the `system_requirements()` method from within `package_info()`, but before doing so, override the `SystemPackageTool()` install method with an alternate implementation (invoked only when `system_requirements()` is invoked from `package_info()`).  Instead of actually running `apt get`, it could store each value it gets called with to an internal list. Then, at the end, that list could be written to a file, and then, that file could be read inside `package_info()` method with some helper function.  Then, the values could be assigned to the new `self.sysrequires_info`.   We could do the same for all the other platform specific tools like `apt()`/`yum()`. 

- The next step, is to seek approval from @memsharded that `package_info()` should even support passing system package names in the conan dependency data model.

**Comment by KerstinKeller** [comment]
Thanks for the very detailed answer.
But I am wondering just one thing: if this info so far is not propagated at all, does it even reflect in the package_id() of a package? I had somehow naively assumed that it would, the same way some other requirement influences the package id.

It's good to agree that this information is useful and necessary, how to manage it and make it available - in that regard you are the experts.

For me it seems natural that that they would somehow be available in the `package_info` method, just like regular dependencies are are vailable there. A `sysrequires_info` seems like a good fit, though then it's probably also need to be accounted for which system package tools it "belongs to".

**Comment by solvingj** [comment]
System requirements are not part of the `package_id` calculation by default.  This is part of the reason why direct use of `system_requirements` is not allowed in most recipes in Conan Center index. It's only allowed in dedicated wrapper recipes/packages which exist purely to encapsulate `system_requirements` of the same name. Then, recipes which would declare a `system_requirement` , instead declare a regular `requirement` on the wrapper package and this indirectly makes the `system_requirement` part of the `package_id` of the consumer.

**Comment by memsharded** [comment]
This has been addressed in Conan 2.0, it is possible to collect system requirements in json output with ``graph info .  --format=json``, it reports and serializes to the json the system-requirements, and it also follows the different ``conf`` that affects the system requirements.

Closing this as solved and implemented in Conan 2.0, please try it and feedback welcome!
